### PR TITLE
Make While Op could run on GPU place and add while_loop unittest 

### PR DESCRIPTION
**Pull request by gfwm2013**
1.更改代码使得While Op可以在GPU上运行
2.添加while_loop的反向单测，while_loop和switch_case混和使用的单测，和while_loop的嵌套网络的反向单测

### CODE REVIEW

**Review by zhhsplendid — COMMENTED** [review]
You could split this into two PRs in the future. It's fine this time.

File: paddle/fluid/operators/controlflow/while_op_helper.h [hunk 1]
```diff
@@ -40,5 +42,21 @@ void PrepareSafeEagerDeletionOnWhileOpAndWhileGradOp(
     const std::vector<framework::OperatorBase *> &while_ops,
     const std::vector<framework::OperatorBase *> &while_grad_ops);
 
+inline bool GetCondData(const framework::LoDTensor &cond) {
```
Comment (zhhsplendid) [inline paddle/fluid/operators/controlflow/while_op_helper.h]: Had better to put declaration at .h file and implementation at .cc file. Not inline it.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 2]
```diff
@@ -110,24 +113,146 @@ def internal_body(j, init, sums):
...
+                x = layers.elementwise_add(x=x, y=one)
+                i = layers.increment(i)
+                return [i, x]
+
+            temp = layers.while_loop(cond_inside, body_inside, [i, x])
+            x = layers.elementwise_add(x=temp[1], y=two)
+            i = layers.increment(i)
+            return [i, x]
+
+        main_program = fluid.default_main_program()
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: To make sure your program in this test is completely empty, would you mind changing those to

main_program = fluid.Program()
startup_program = fluid.Program()

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 3]
```diff
@@ -110,24 +113,146 @@ def internal_body(j, init, sums):
...
+
+            def fn_3():
+                data_add_one = layers.elementwise_add(x=i, y=one)
+                return data_add_one
+
+            return layers.switch_case(
+                branch_index=i, branch_fns={2: fn_1,
+                                            5: fn_2}, default=fn_3)
+
+        main_program = fluid.default_main_program()
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: To make sure your program in this test is completely empty, would you mind changing those to

main_program = fluid.Program()
startup_program = fluid.Program()

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 4]
```diff
@@ -110,24 +113,146 @@ def internal_body(j, init, sums):
...
+        self.assertTrue(np.allclose(np.asarray(res[1]), x_grad))
+
+
+class TestApiWhileLoopWithSwitchCase(unittest.TestCase):
+    def test_with_(self):
+        def cond(i):
+            return layers.less_than(i, ten)
+
+        def body(i):
+            def fn_1():
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Best name for fn_1,2,3 may be fn_add_three, fn_square, fn_add_one

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 5]
```diff
@@ -110,24 +113,146 @@ def internal_body(j, init, sums):
...
+                      fetch_list=[mean.name, mean.grad_name])
+
+        feed_x = np.asarray([16]).astype("float32")
+        x_grad = np.ones(1).astype("int32")
+        self.assertTrue(np.allclose(np.asarray(res[0]), feed_x))
+        self.assertTrue(np.allclose(np.asarray(res[1]), x_grad))
+
+
+class TestApiWhileLoopWithSwitchCase(unittest.TestCase):
+    def test_with_(self):
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: This name "test_with_" confused me. You may change it into "test_with_switch_case"?

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 6]
```diff
@@ -110,24 +113,146 @@ def internal_body(j, init, sums):
...
+
+            mean = layers.mean(out[1])
+            append_backward(mean)
+
+        place = fluid.CUDAPlace(0) if core.is_compiled_with_cuda(
+        ) else fluid.CPUPlace()
+        exe = fluid.Executor(place)
+        res = exe.run(main_program,
+                      feed={},
+                      fetch_list=[mean.name, mean.grad_name])
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: I saw you fetched mean and mean.grad_name but compare it with feed_x and x_grad, is it an error? If yes, please change them, compare the grad value of x or mean value.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 7]
```diff
@@ -110,24 +113,146 @@ def internal_body(j, init, sums):
...
+            out = layers.while_loop(cond, body, [i, x])
+            mean = layers.mean(out[1])
+            append_backward(mean)
+
+        place = fluid.CUDAPlace(0) if core.is_compiled_with_cuda(
+        ) else fluid.CPUPlace()
+        exe = fluid.Executor(place)
+
+        feed_x = np.asarray([10]).astype('float32')
+        res = exe.run(program, feed={}, fetch_list=[mean.name, mean.grad_name])
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: I saw you fetched mean and mean.grad_name but compare it with feed_x and x_grad, is it an error? If yes, please change them, compare the grad value of x or mean value.

File: paddle/fluid/operators/controlflow/while_op_helper.h (hunk 1 above)
Comment (gfwm2013) [reply paddle/fluid/operators/controlflow/while_op_helper.h]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py (hunk 2 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py (hunk 3 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py (hunk 4 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py (hunk 5 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py (hunk 7 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py (hunk 6 above)
Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

File: python/paddle/fluid/tests/unittests/test_while_loop_op.py [hunk 8]
```diff
@@ -104,30 +111,183 @@ def internal_body(j, init, sums):
...
+            i = layers.increment(i)
+            return [i, x]
+
+        main_program = Program()
+        startup_program = Program()
+        with fluid.program_guard(main_program, startup_program):
+            i = layers.data(
+                name='i', shape=[1], dtype='float32', append_batch_size=False)
+            i.stop_gradient = False
+            ten = layers.fill_constant(shape=[1], dtype='float32', value=11)
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Change the "ten" to "eleven"

Comment (gfwm2013) [reply python/paddle/fluid/tests/unittests/test_while_loop_op.py]: Done.

**Review by zhhsplendid — DISMISSED** [review]
LGTM

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**